### ISSUE TITLE
Type inference failure resulting in spurious errors in generic function using IntoIterator

### ISSUE THREAD

**Issue by axnsan12** [intellij-rust/intellij-rust#10111]
## Environment

* **IntelliJ Rust plugin version:** 0.4.187.5175-223
* **Rust toolchain version:** 1.68.0-nightly
* **IDE name and version:** CLion 2022.3.2
* **Operating system:** Windows 10

## Problem description

```rust
fn test(inputs: impl IntoIterator<Item = Option<u32>>) -> Option<u32> {
    let iter = inputs.into_iter();
    let ret = iter.reduce(|acc, current| {
        match (acc, current) {
            (Some(acc), Some(current)) => Some(acc + current),
            _ => None,
        }
    });

    ret.flatten()
}
```

Shows 

```
mismatched types [E0308] expected `Option<u32>`, found `Flatten<&mut Option<<<impl IntoIterator<Item=Option<u32>, IntoIter=<unknown>>+Sized as IntoIterator>::IntoIter as Iterator>::Item>>` 
``` 

in the IDE on the last line.

![image](https://user-images.githubusercontent.com/6382979/218172312-3d934b0f-e4e5-4b8b-a19e-fac6468ad4d6.png)

There are two things that can make the error go away:
- change the argument type from `impl IntoIterator<Item = Option<u32>>` to `Vec<Option<u32>>`
- annotate the `ret` variable with `let ret: Option<Option<u32>> = ..`

## Steps to reproduce

Write the function above into a .rs file.
